pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -0,0 +1,405 @@
+    Map<String, Map<String, String>> nextAssignment = new TreeMap<>();
+
+    for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
+      String segmentName = entry.getKey();
+      nextAssignment.put(segmentName,
+          getNextInstanceStateMap(entry.getValue(), targetAssignment.get(segmentName), minAvailableReplicas));
+    }
+
+    return nextAssignment;
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("Duplicates")
+  static Map<String, String> getNextInstanceStateMap(Map<String, String> currentInstanceStateMap,
+      Map<String, String> targetInstanceStateMap, int minAvailableReplicas) {
+    Map<String, String> nextInstanceStateMap = new TreeMap<>();
+
+    // Add all the common instances
+    for (Map.Entry<String, String> entry : targetInstanceStateMap.entrySet()) {
+      String instanceName = entry.getKey();
+      if (currentInstanceStateMap.containsKey(instanceName)) {
+        nextInstanceStateMap.put(instanceName, entry.getValue());
+      }
+    }
+
+    // Add current instances until the min available replicas achieved
+    int instancesToKeep = minAvailableReplicas - nextInstanceStateMap.size();
+    if (instancesToKeep > 0) {
+      for (Map.Entry<String, String> entry : currentInstanceStateMap.entrySet()) {
+        String instanceName = entry.getKey();
+        if (!nextInstanceStateMap.containsKey(instanceName)) {
+          nextInstanceStateMap.put(instanceName, entry.getValue());
+          if (--instancesToKeep == 0) {
+            break;
+          }
+        }
+      }
+    }
+
+    // Add target instances until the number of instances matched
+    int instancesToAdd = targetInstanceStateMap.size() - nextInstanceStateMap.size();
+    if (instancesToAdd > 0) {
+      for (Map.Entry<String, String> entry : targetInstanceStateMap.entrySet()) {
+        String instanceName = entry.getKey();
+        if (!nextInstanceStateMap.containsKey(instanceName)) {
+          nextInstanceStateMap.put(instanceName, entry.getValue());
+          if (--instancesToAdd == 0) {
+            break;
+          }
+        }
+      }
+    }
+
+    return nextInstanceStateMap;
+  }
+
+  private static boolean hasSegmentInErrorState(Map<String, Map<String, String>> segmentStates) {
+    for (Map<String, String> instanceStateMap : segmentStates.values()) {
+      if (instanceStateMap.containsValue(SegmentOnlineOfflineStateModel.ERROR)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * NOTE: Only check the instances in IdealState. When dropping the segment, Helix might schedule multiple
+   * ONLINE->OFFLINE state transitions, and could leave segment in OFFLINE state instead of dropping the segment. In
+   * such case we don't want to block forever. It is okay to have not match instance state map as long as all the states
+   * in IdealState are reached.
+   */
+  @VisibleForTesting
+  static boolean isExternalViewConverged(Map<String, Map<String, String>> externalViewSegmentStates,
```

Comment:
Are we no longer checking if external view has reached error state (and if so it may never converge and thus rebalancer will go on and on indefinitely?)

Reply:
We check that with method TableRebalancer.hasSegmentInErrorState()